### PR TITLE
Fix some copying issues.

### PR DESCRIPTION
**Pull request by WeatherGod**
Issues discovered while working on the key rename PR.

### CODE REVIEW

File: cycler.py [hunk 1]
```diff
@@ -113,10 +113,26 @@ def __init__(self, left, right=None, op=None):
         Do not use this directly, use `cycler` function instead.
         """
         self._keys = _process_keys(left, right)
-        self._left = copy.deepcopy(left)
-        self._right = copy.deepcopy(right)
+        if isinstance(left, Cycler):
+            self._left = left._shallow_copy()
+        else:
+            self._left = copy.copy(left)
```
Comment (tacaswell) [inline cycler.py]: This isn't deep enough to protect against changing the key, but this depends on how #13 ends up being implemented

```
In [80]: dd = [{a: a} for a in range(3)]

In [81]: import copy

In [82]: dd2 = copy.copy(dd)

In [83]: dd[0][0] = 'foobare'

In [84]: dd2
Out[84]: [{0: 'foobare'}, {1: 1}, {2: 2}]
```

I think this needs to be 

```python
self._left = [dict(v) for v in left]

File: cycler.py [hunk 2]
```diff
@@ -113,10 +113,26 @@ def __init__(self, left, right=None, op=None):
...
+
+        if isinstance(right, Cycler):
+            self._right = right._shallow_copy()
+        else:
+            self._right = copy.copy(right)
+
         self._op = op
 
+    def _shallow_copy(self):
+        ret = Cycler(None)
```
Comment (tacaswell) [inline cycler.py]: I think we can do this block as `return Cycler(self._left, self._right, self.op)` 

ps, sorry for the  backwards lisp

Comment (WeatherGod) [inline cycler.py]: we would need to expand the constructor to allow lists of dictionaries and update _process_keys() to recognize that.

**Comment by WeatherGod** [comment]
ok, I reworked it a bit, and even managed to get rid of the ugly `_shallow_copy()` function.

**Comment by tacaswell** [comment]
Oh, that is nice.  If I had thought of that originally the function `cycler` would not exist.

**Pull request merged**